WIP: Kopano Konnect und Rest API

fbartels opened the pull request.

fbartels committed:
add konnect to compose file

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
konnect now succesfully starts

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
adapt gencerts for konnect certs

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
integrate gencerts into start.sh
adapt if to only skip individual parts

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add a new image for konnect
add certificates of konnect in ssl image

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add container for kapi

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
use same config for webapp and z-push as https://stash.kopano.io/projects/KGOL/repos/kweb/browse/config/legacy.go

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
users can now login via konnect

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
fix merge mistakes setup.sh

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
grapi starts, but is using the wrong socket

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
getting closer. only failing at `kcoid discovery error: Get https://kopano.demo/.well-known/openid-configuration: dial tcp: lookup kopano.demo on 127.0.0.11:53: no such host`
`

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
now it can connect, but
`kcoid discovery error: Get https://kopano.demo/.well-known/openid-configuration: dial tcp 192.168.64.5:443: connect: connection refused`

Signed-off-by: Felix Bartels <felix@host-consultants.de>

longsleep commented on konnect/wrapper.sh:
```diff
@@ -0,0 +1,14 @@
+	-wait file:///kopano/ssl/konnectd-encryption.key \
+	-timeout 360s \
+	sleep 2
+exec konnectd serve \
```
That seems unclean - why dockerize sleep?

fbartels replied:
Sleep could also be at a dedicated line and not be part of the dockerize command. The problem where the sleep helps is that dockerize continues once the file has appeared. So even though the file is still being written. That in some cases lead to konnect failing to start since the certificate wasn't valid/complete.

longsleep replied:
Mhm certificates and keys should be created atomic. Means they write to a temp file and once complete move the full file to the target which avoids half write issues. How are those files created? Needs to be fixed there instead of here.

fbartels replied:
yes, that is indeed the better way to solve this. implemented

longsleep commented on core/start-service.sh:
```diff
@@ -72,6 +72,30 @@ ical)
 	;;
+grapi)
+	dockerize \
+		-wait tcp://kopano_server:236 \
```
There is no need to wait on server.

longsleep commented on docker-compose.yml-example:
```diff
@@ -3,7 +3,7 @@ version: "3.5"
   web:
     image: ${docker_repo:?err}/kopano_web
-    container_name: web
+    container_name: ${FQDN}
```
Imho this will only cause issues and makes it impossible to resolve the fqdn externally - strongly suggest to keep this web.

fbartels committed:
use example compose file in make test target
remove dockerize from grapi startup
switch name of web container back
let kapi connect directly to konnect
add comment to sleep in konnect startup

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
remove legacy links

Signed-off-by: Felix Bartels <felix@host-consultants.de>

longsleep commented on Makefile:
```diff
@@ -137,13 +140,13 @@ publish-kweb: build-kweb
 
 test:
-	docker-compose down -v || true
+	docker-compose -f docker-compose.yml-example down -v || true
```
Maybe add the name of the compose file as an overridable make variable ?

longsleep commented on core/start-service.sh:
```diff
@@ -72,6 +72,27 @@ ical)
+	chown -R kapi:kopano $socket_path
+	# cleaning up env variables
+	unset "${!KCCONF_@}"
+	exec /usr/sbin/kopano-grapi serve
```
It would be nice for all exec lines to use no path - makes it much better to read in the process list and the bin sbin paths should be in the system path list anyways.

fbartels replied:
good point

longsleep commented on docker-compose.yml-example:
```diff
@@ -215,18 +201,42 @@ services:
+      - KCCONF_KAPID_OIDC_ISSUER_IDENTIFIER=http://kopano_konnect:8777 # if this works, then it should go in the python script
+    networks:
+      - kopano-net
+      - web-net
```
Maybe consider renaming those to both have the kopano prefix - why are there two networks anyways?

fbartels replied:
There is even a third one for ldap. its probably a bit on the paranoid side, but the suggested idea was to have only these containers in the same network, that need to see each other.

as for the prefix, this is already done when the network is through compose:
```
23eb696245db        kopano-docker_kopano-net   bridge              local
21a53ce5d83f        kopano-docker_web-net      bridge              local
```

longsleep replied:
ok got - seems a bit complicated for minimal gain but ok

longsleep commented on ssl/start.sh:
```diff
@@ -1,7 +1,40 @@
 
-if [ -f /kopano/ssl/ca.pem ]; then
-	exit 0
+if [ ! -f /kopano/ssl/ca.pem ]; then
```
For all the files which get checked in conditions make sure they are created with atomic operations like mv or ln to avoid half writes and incomplete states. Also ensure that the atomic operation is last. Means always write to tmp files or new files and move or link when all successfully written.

fbartels replied:
good point. i have adapted the script.

fbartels committed:
write certificates first to a tempfile

resolves https://github.com/zokradonh/kopano-docker/pull/72#discussion_r245485815

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
fix certificate creation
make compose file configurable through an env variable

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
remove sleep from konnect wrapper

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
remove unneccesary paths

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
make all ssl creation atomic

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
move copying of ca into last line

Signed-off-by: Felix Bartels <felix@host-consultants.de>

longsleep commented on docker-compose.yml-example:
```diff
@@ -215,18 +201,42 @@ services:
+      - SERVICE_TO_START=kapid
+      - TZ=${TZ}
+      - KCCONF_KAPID_LOG_LEVEL=DEBUG
+      - KCCONF_KAPID_OIDC_ISSUER_IDENTIFIER=http://kopano_konnect:8777 # if this works, then it should go in the python script
```
While this works to load the discovery document, it ultimatively still will not work when the iss as configured within konnect is not resolvable by kapid since the discovery document contains links which need to be loaded (jwks for example).

fbartels replied:
Yes, I have already realized that. This will need to go back to $fqdn.

longsleep commented on ssl/start.sh:
```diff
@@ -1,7 +1,46 @@
+		cat /kopano/easypki/internalca/certs/$s.crt >> /tmp/$s.pem
+		openssl x509 -in /kopano/easypki/internalca/certs/$s.crt -pubkey -noout > /tmp/$s-public.pem
+		cp /tmp/$s.pem /kopano/ssl/$s.pem
+		cp /tmp/$s-public.pem /kopano/ssl/clients/$s-public.pem
```
cp is not atomic while rm has the chance to be atomic. So also the choice of location matters. Do not use tmp, use a tmp file in the same folder as the destination to write to and after having created the file, move it to the desired name in the same folder. So no /tmp, no cp.

longsleep replied:
This is for all the create stuff also in other files and down in this script. (Since mv is not atomic when crossing file systems and /tmp is most certainly a different file system like tmpfs).

fbartels replied:
ok, misconception from my side. now files are created in their location with a tmp file extension and moved afterwards.

fbartels committed:
add option to run kapi insecure for testing
currently kapi is starting to fail because of missing database migration

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
really make ssl tasks atomic

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
adapt needed paths

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
configure openid for kopano-server

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
kapi playground

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add example for konnect device registry

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add local playground to test functionality of konnect and kapi

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
print errors to stdout as well

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add extrahosts to kopano_server

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
adapt extrahosts for ip command on ubuntu 18.04

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
environment variables need to be uppercase

Signed-off-by: Felix Bartels <felix@host-consultants.de>

fbartels committed:
add documentation

Signed-off-by: Felix Bartels <felix@host-consultants.de>
